##Dataset 1 — Appliances Energy Prediction (UCI)
Situação

fonte: https://archive.ics.uci.edu/dataset/374/appliances+energy+prediction

Uma empresa de eficiência energética está analisando o comportamento de uma residência de baixo consumo. A equipe deseja identificar períodos de consumo elevado dos eletrodomésticos e observar quais condições de temperatura e umidade estavam presentes nesses momentos.


In [ ]:
# 1.	Carregue a amostra no Pandas e apresente head(), shape, info() e describe().
import pandas as pd

dados = pd.read_csv("energydata_sample.csv")
dados.head()




,Appliances,lights,T1,RH_1,T2,RH_2,T3,RH_3
0,40,0,20.8900,35.4000,17.7600,39.1633,20.2900,36.9000
1,90,10,21.8900,53.1000,21.2900,45.3600,21.6333,49.2267
2,50,0,21.3900,35.5000,17.6333,40.5300,21.6667,35.2000
3,50,0,21.3900,41.0333,23.8900,34.8400,22.0333,36.9333
4,70,0,19.9633,35.1267,16.4633,40.1267,20.0000,36.4000


In [ ]:
dados.shape

(1974, 8)

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1974 entries, 0 to 1973
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Appliances  1974 non-null   int64  
 1   lights      1974 non-null   int64  
 2   T1          1974 non-null   float64
 3   RH_1        1974 non-null   float64
 4   T2          1974 non-null   float64
 5   RH_2        1974 non-null   float64
 6   T3          1974 non-null   float64
 7   RH_3        1974 non-null   float64
dtypes: float64(6), int64(2)
memory usage: 123.5 KB


In [ ]:
dados.describe()

,Appliances,lights,T1,RH_1,T2,RH_2,T3,RH_3
count,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000
mean,93.044580,3.839919,21.664703,40.219193,20.293635,40.411386,22.231615,39.209733
std,93.826149,7.937076,1.574883,3.971923,2.126958,4.037686,1.957275,3.288818
min,20.000000,0.000000,16.790000,27.733300,16.100000,24.063300,17.200000,30.133300
25%,50.000000,0.000000,20.700000,37.205825,18.790000,37.760000,20.700000,36.900000
50%,60.000000,0.000000,21.600000,39.590000,20.000000,40.433300,22.100000,38.433300
75%,100.000000,0.000000,22.600000,42.991675,21.500000,43.228325,23.290000,41.741250
max,770.000000,50.000000,26.260000,57.496700,28.917500,56.026700,29.198600,49.226700


In [ ]:
# 2.	Renomeie Appliances para Consumo_Eletrodomesticos e simplifique o nome de pelo menos três atributos ambientais.
dados.rename(columns={'Appliances': 'Consumo_Eletrodomesticos'}, inplace=True)
dados.rename(columns={'T1': 'Temperatura_Setor_1'}, inplace=True)
dados.rename(columns={'T2': 'Temperatura_Setor_2'}, inplace=True)
dados.rename(columns={'T3': 'Temperatura_Setor_3'}, inplace=True)

dados.head()

,Consumo_Eletrodomesticos,lights,Temperatura_Setor_1,RH_1,Temperatura_Setor_2,RH_2,Temperatura_Setor_3,RH_3
0,40,0,20.8900,35.4000,17.7600,39.1633,20.2900,36.9000
1,90,10,21.8900,53.1000,21.2900,45.3600,21.6333,49.2267
2,50,0,21.3900,35.5000,17.6333,40.5300,21.6667,35.2000
3,50,0,21.3900,41.0333,23.8900,34.8400,22.0333,36.9333
4,70,0,19.9633,35.1267,16.4633,40.1267,20.0000,36.4000


In [ ]:
# 3.	Determine o maior consumo de eletrodomésticos registrado na amostra.
dados['Consumo_Eletrodomesticos'].max()
dmax = dados['Consumo_Eletrodomesticos'].max()


In [ ]:
# 4.	Calcule um limiar correspondente a 70% do valor máximo e crie um DataFrame contendo somente os registros acima desse limite.
c70 = dmax*0.7
print(f"O limiar 70% é: {c70}")


O limiar 70% é: 539.0


In [ ]:
df70 = dados[dados['Consumo_Eletrodomesticos'] > c70]
df70.head()

,Consumo_Eletrodomesticos,lights,Temperatura_Setor_1,RH_1,Temperatura_Setor_2,RH_2,Temperatura_Setor_3,RH_3
352,620,0,24.3900,44.3333,25.3700,37.7360,26.7300,38.8633
427,650,0,20.8900,36.7000,18.2900,39.6633,21.2000,36.0000
453,560,30,21.3567,36.0900,19.5667,37.1633,21.6333,35.0300
616,570,10,24.8900,32.6667,23.5600,31.1400,23.6000,30.8233
696,590,30,19.8567,47.6633,18.7000,34.1000,18.4267,37.8333


In [ ]:
# 5.	Conte os registros selecionados e calcule qual percentual da amostra eles representam.
num_registros_acima_limite = df70.shape[0]
percentual_registros_acima_limite = (num_registros_acima_limite / dados.shape[0]) * 100

print(f"Número de registros com Consumo_Eletrodomesticos acima do limiar de 70%: {num_registros_acima_limite}")
print(f"Percentual da amostra que esses registros representam: {percentual_registros_acima_limite:.2f}%")

Número de registros com Consumo_Eletrodomesticos acima do limiar de 70%: 22
Percentual da amostra que esses registros representam: 1.11%


In [ ]:
# 6.	Calcule a temperatura média de T1 e crie outro DataFrame contendo simultaneamente consumo acima de 70% do máximo e temperatura acima da média.
media_t1 = dados['Temperatura_Setor_1'].mean()
print(f"A temperatura média de Temperatura_Setor_1 é: {media_t1:.2f}")

A temperatura média de Temperatura_Setor_1 é: 21.66


In [ ]:
df_consumo_temperatura_alta = dados[(dados['Consumo_Eletrodomesticos'] > c70) & (dados['Temperatura_Setor_1'] > media_t1)]
display(df_consumo_temperatura_alta.head())

,Consumo_Eletrodomesticos,lights,Temperatura_Setor_1,RH_1,Temperatura_Setor_2,RH_2,Temperatura_Setor_3,RH_3
352,620,0,24.3900,44.3333,25.3700,37.7360,26.73,38.8633
616,570,10,24.8900,32.6667,23.5600,31.1400,23.60,30.8233
916,750,30,22.4300,43.0000,21.9267,40.1333,21.23,42.5967
993,690,10,22.5333,45.0267,22.0667,39.7900,22.39,41.9300
1182,600,20,24.8900,48.1333,28.9175,37.9475,27.10,42.4400


#### 7.	Compare os dois DataFrames e explique o efeito da inclusão da temperatura como segundo critério.



In [ ]:
num_registros_duplo_criterio = df_consumo_temperatura_alta.shape[0]
percentual_duplo_criterio = (num_registros_duplo_criterio / dados.shape[0]) * 100

print(f"Número de registros com Consumo_Eletrodomesticos acima do limiar de 70% E Temperatura_Setor_1 acima da média: {num_registros_duplo_criterio}")
print(f"Percentual da amostra que esses registros representam (com duplo critério): {percentual_duplo_criterio:.2f}%")

Número de registros com Consumo_Eletrodomesticos acima do limiar de 70% E Temperatura_Setor_1 acima da média: 8
Percentual da amostra que esses registros representam (com duplo critério): 0.41%


Comparando os dois DataFrames:

**`df70` (apenas consumo alto):** Contém **22 registros**, representando **1.11%** da amostra total.
**`df_consumo_temperatura_alta` (consumo alto E temperatura acima da média):** Contém um número menor de registros, o que indica que, ao adicionar o critério de temperatura, estamos refinando a seleção e identificando um subconjunto mais específico de eventos de alto consumo.

O efeito da inclusão da temperatura como segundo critério é a **redução do número de registros** que atendem às condições. Isso significa que, nem todos os períodos de alto consumo de eletrodomésticos ocorrem quando a `Temperatura_Setor_1` está acima da média. Ao adicionar o critério de temperatura, isolamos os momentos de alto consumo que são *simultaneamente* acompanhados por temperaturas mais elevadas no Setor 1. Isso pode ser útil para identificar padrões específicos de comportamento ou condições que levam a picos de consumo.

##Dataset 2 — Steel Industry Energy Consumption (UCI)

fonte: https://archive.ics.uci.edu/dataset/851/steel+industry+energy+consumption

Situação
A equipe de gestão de energia de uma indústria siderúrgica deseja localizar situações de consumo elevado e verificar se esses períodos coincidem com condições de carga elevada ou valores menos favoráveis de fator de potência.


In [ ]:
# 1.	Carregue a amostra e renomeie Usage_kWh para Consumo_kWh, simplificando também os nomes dos atributos de fator de potência.
dados = pd.read_csv('Steel_industry_sample.csv')
dados.head()

,date,Usage_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Lagging_Current_Power_Factor,Leading_Current_Power_Factor,WeekStatus,Day_of_week,Load_Type
0,29/04/2018 07:15,2.88,3.82,0.0,60.20,100.00,Weekend,Sunday,Light_Load
1,04/10/2018 12:00,60.77,48.02,0.0,78.46,100.00,Weekday,Thursday,Maximum_Load
2,26/01/2018 11:30,120.42,59.65,0.0,89.61,100.00,Weekday,Friday,Maximum_Load
3,02/06/2018 14:30,3.13,0.00,16.6,100.00,18.53,Weekend,Saturday,Light_Load
4,07/12/2018 15:00,58.86,20.99,0.0,94.19,100.00,Weekday,Friday,Medium_Load


In [ ]:

dados.rename(columns={'Usage_kWh': 'Consumo_kWh'}, inplace=True)
dados.rename(columns={'Lagging_Current_Reactive.Power_kVarh': 'RP_X'}, inplace=True)
dados.rename(columns={'Leading_Current_Reactive_Power_kVarh': 'RP_Y'}, inplace=True)
dados.rename(columns={'Lagging_Current_Power_Factor': 'PF_X'}, inplace=True)
dados.rename(columns={'Leading_Current_Power_Factor': 'PF_Y'}, inplace=True)

In [ ]:
# 2.	Apresente uma inspeção inicial com head(), shape, info() e describe().
dados.head()

,date,Consumo_kWh,RP_X,RP_Y,PF_X,PF_Y,WeekStatus,Day_of_week,Load_Type
0,29/04/2018 07:15,2.88,3.82,0.0,60.20,100.00,Weekend,Sunday,Light_Load
1,04/10/2018 12:00,60.77,48.02,0.0,78.46,100.00,Weekday,Thursday,Maximum_Load
2,26/01/2018 11:30,120.42,59.65,0.0,89.61,100.00,Weekday,Friday,Maximum_Load
3,02/06/2018 14:30,3.13,0.00,16.6,100.00,18.53,Weekend,Saturday,Light_Load
4,07/12/2018 15:00,58.86,20.99,0.0,94.19,100.00,Weekday,Friday,Medium_Load


In [ ]:
dados.shape

(7008, 9)

In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7008 entries, 0 to 7007
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         7008 non-null   object 
 1   Consumo_kWh  7008 non-null   float64
 2   RP_X         7008 non-null   float64
 3   RP_Y         7008 non-null   float64
 4   PF_X         7008 non-null   float64
 5   PF_Y         7008 non-null   float64
 6   WeekStatus   7008 non-null   object 
 7   Day_of_week  7008 non-null   object 
 8   Load_Type    7008 non-null   object 
dtypes: float64(5), object(4)
memory usage: 492.9+ KB


In [ ]:
dados.describe()

,Consumo_kWh,RP_X,RP_Y,PF_X,PF_Y
count,7008.000000,7008.000000,7008.000000,7008.000000,7008.000000
mean,27.790164,13.206839,3.803711,80.672277,84.491448
std,33.717719,16.444246,7.356412,18.869855,30.314798
min,2.450000,0.000000,0.000000,37.870000,12.710000
25%,3.200000,2.380000,0.000000,63.602500,99.710000
50%,4.570000,5.040000,0.000000,87.990000,100.000000
75%,52.107500,22.977500,1.857500,98.842500,100.000000
max,153.140000,87.700000,27.650000,100.000000,100.000000


In [ ]:
# 3.	Determine o maior consumo registrado e calcule um limiar de 75% do consumo máximo.
dados['Consumo_kWh'].max()
dmax = dados['Consumo_kWh'].max()
df75 = dmax*0.75
print(f"O limiar máximo é: {df75:.3f}")



O limiar máximo é: 114.855


In [ ]:
# 4.	Crie um DataFrame contendo os registros acima desse limiar e calcule a quantidade e o percentual de registros.
df75 = dados[dados['Consumo_kWh'] > df75]
df75.head()

,date,Consumo_kWh,RP_X,RP_Y,PF_X,PF_Y,WeekStatus,Day_of_week,Load_Type
2,26/01/2018 11:30,120.42,59.65,0.0,89.61,100.0,Weekday,Friday,Maximum_Load
62,08/02/2018 14:30,118.22,54.04,0.0,90.95,100.0,Weekday,Thursday,Medium_Load
100,29/01/2018 16:45,129.13,46.58,0.0,94.07,100.0,Weekday,Monday,Medium_Load
108,29/01/2018 14:00,117.07,46.44,0.0,92.95,100.0,Weekday,Monday,Medium_Load
159,22/11/2018 10:15,139.57,58.90,0.0,92.13,100.0,Weekday,Thursday,Maximum_Load


In [ ]:
num_registros_acima_limite = df75.shape[0]
percentual_registros_acima_limite = (num_registros_acima_limite / dados.shape[0]) * 100

print(f"Número de registros com Consumo_kWh acima do limiar de 75%: {num_registros_acima_limite}")
print(f"Percentual da amostra que esses registros representam: {percentual_registros_acima_limite:.2f}%")

Número de registros com Consumo_kWh acima do limiar de 75%: 129
Percentual da amostra que esses registros representam: 1.84%


In [ ]:
# 5.	Verifique quantos registros desse DataFrame pertencem à categoria Maximum Load.
registros_maximum_load = df75[df75['Load_Type'] == 'Maximum_Load'].shape[0]
print(f"Número de registros na categoria 'Maximum_Load' dentro do df75: {registros_maximum_load}")

Número de registros na categoria 'Maximum_Load' dentro do df75: 62


In [ ]:
# 6.	Escolha um dos atributos de fator de potência, observe seus valores e defina um limite coerente para representar valores mais baixos.
limite_pf_baixo = dados['PF_X'].quantile(0.25)
print(f"Atributo escolhido para fator de potência: PF_X")
print(f"O 25º percentil (limite para valores mais baixos) de PF_X é: {limite_pf_baixo:.2f}")

Atributo escolhido para fator de potência: PF_X
O 25º percentil (limite para valores mais baixos) de PF_X é: 63.60


In [ ]:
# 7. Crie um novo DataFrame contendo simultaneamente consumo elevado e fator de potência abaixo do limite escolhido.
consumo_limiar_alto = dmax * 0.75
df_consumo_pf_baixo = dados[(dados['Consumo_kWh'] > consumo_limiar_alto) | (dados['PF_X'] < limite_pf_baixo)]
display(df_consumo_pf_baixo.head())

,date,Consumo_kWh,RP_X,RP_Y,PF_X,PF_Y,WeekStatus,Day_of_week,Load_Type
0,29/04/2018 07:15,2.88,3.82,0.0,60.20,100.0,Weekend,Sunday,Light_Load
2,26/01/2018 11:30,120.42,59.65,0.0,89.61,100.0,Weekday,Friday,Maximum_Load
10,15/07/2018 04:30,2.84,5.26,0.0,47.51,100.0,Weekend,Sunday,Light_Load
11,23/05/2018 06:00,3.02,5.26,0.0,49.79,100.0,Weekday,Wednesday,Light_Load
12,26/07/2018 23:15,2.84,5.36,0.0,46.82,100.0,Weekday,Thursday,Light_Load


#### 8.	Explique por que esse segundo conjunto de registros pode merecer maior atenção da equipe de energia.



A equipe de energia deve dar maior atenção a este segundo conjunto de registros (`df_consumo_pf_baixo`) porque eles representam períodos em que a **demanda de energia é alta (consumo elevado)** e, simultaneamente, o **fator de potência está baixo**. Esta combinação é particularmente problemática por várias razões:

1.  **Ineficiência Energética e Perdas:** Um baixo fator de potência indica que a energia elétrica está sendo usada de forma ineficiente. Parte da energia reativa (que não realiza trabalho útil) aumenta as perdas na rede de distribuição e nos equipamentos da indústria. Isso significa que, para o mesmo trabalho útil, a empresa está consumindo mais energia total, resultando em desperdício.
2.  **Multas e Custos Adicionais:** Empresas distribuidoras de energia frequentemente aplicam multas a consumidores que operam com fator de potência abaixo de um limite regulatório (geralmente próximo de 0.92 ou 0.95 no Brasil). Identificar e corrigir esses períodos pode evitar custos extras significativos na fatura de energia.
3.  **Sobrecarga de Equipamentos:** Um baixo fator de potência exige que a rede elétrica e os equipamentos internos (transformadores, cabos, geradores) transportem uma corrente maior para fornecer a mesma quantidade de energia ativa. Isso pode levar a superaquecimento, desgaste prematuro dos equipamentos e redução de sua vida útil, além de exigir investimentos em infraestrutura mais robusta.
4.  **Instabilidade do Sistema:** Períodos prolongados de baixo fator de potência em conjunto com alto consumo podem contribuir para a instabilidade da rede elétrica local, afetando a qualidade da energia fornecida e, em casos extremos, causando quedas de tensão ou interrupções.

Portar, esses registros são cruciais para identificar oportunidades de **otimização da qualidade da energia**, **redução de custos operacionais** e **aumento da confiabilidade e vida útil dos equipamentos** da indústria siderúrgica. A equipe pode investigar as causas específicas desses eventos (e.g., operação de motores indutivos sem compensação, uso de equipamentos antigos) e implementar soluções como bancos de capacitores para correção do fator de potência.

## Dataset 3 — Power Consumption of Tetouan City (UCI)

fonte: https://archive.ics.uci.edu/dataset/849/power+consumption+of+tetouan+city

Situação
Uma equipe responsável pelo planejamento da distribuição de energia em Tétouan deseja identificar qual das três zonas apresenta o maior pico de consumo e observar as condições ambientais registradas nos momentos de maior demanda.


In [ ]:
# 1.	Renomeie as três variáveis de consumo para Consumo_Zona_1, Consumo_Zona_2 e Consumo_Zona_3.
dados = pd.read_csv('Tetuan_City_Sample.csv')
dados.rename(columns={'Zone 1 Power Consumption': 'Consumo_Zona_1'}, inplace=True)
dados.rename(columns={'Zone 2  Power Consumption': 'Consumo_Zona_2'}, inplace=True)
dados.rename(columns={'Zone 3  Power Consumption': 'Consumo_Zona_3'}, inplace=True)

dados.head()

,DateTime,Consumo_Zona_1,Consumo_Zona_2,Consumo_Zona_3,Temperature,Humidity,Wind Speed
0,4/18/2017 16:20,32985.14532,19396.7,17623.3,22.76,29.73,0.085
1,12/27/2017 17:10,34737.64259,29803.0,14417.3,15.92,54.44,0.085
2,1/2/2017 11:20,27894.68354,20808.5,16458.8,15.08,57.23,0.076
3,3/19/2017 20:20,42476.93617,23634.1,25089.7,12.42,72.30,0.081
4,4/24/2017 5:50,21341.14101,12006.1,11531.6,14.58,80.60,0.080


In [ ]:
# 2.	Determine o consumo máximo registrado em cada uma das três zonas.
dmax = dados['Consumo_Zona_1'].max()
dmax2 = dados['Consumo_Zona_2'].max()
dmax3 = dados['Consumo_Zona_3'].max()

In [ ]:
# 3.	Identifique qual zona apresenta o maior pico de consumo na amostra.
if dmax > dmax2 and dmax > dmax3:
    print(f'A Zona 1 apresenta o maior pico de consumo: {dmax:.3f}')
    tdmax = dmax
elif dmax2 > dmax and dmax2 > dmax3:
    print(f'A Zona 2 apresenta o maior pico de consumo: {dmax2:.3f}')
    tdmax = dmax2
else:
    print(f'A Zona 3 apresenta o maior pico de consumo: {dmax3:.3f}')
    tdmax = dmax3

A Zona 1 apresenta o maior pico de consumo: 51571.498


In [ ]:
# 4.	Para a zona identificada, calcule 70% do valor máximo e crie um DataFrame com os registros acima desse limiar.
d70 = tdmax*0.70
df70 = dados[dados['Consumo_Zona_1'] > d70]
df70.head()


,DateTime,Consumo_Zona_1,Consumo_Zona_2,Consumo_Zona_3,Temperature,Humidity,Wind Speed
3,3/19/2017 20:20,42476.93617,23634.1,25089.7,12.42,72.30,0.081
8,12/10/2017 17:50,37286.69202,30716.2,15834.8,14.66,65.92,0.075
12,6/5/2017 23:30,40993.90728,25383.4,24960.0,21.03,86.50,0.068
14,8/23/2017 17:50,37475.16093,24276.2,25140.6,27.06,47.58,4.905
19,6/22/2017 2:40,40676.02649,24754.7,25095.9,19.20,65.81,4.924


In [ ]:
# 5.	Conte os registros e calcule o percentual que eles representam.
num_registros_acima_limite = df70.shape[0]
percentual_registros_acima_limite = (num_registros_acima_limite / dados.shape[0]) * 100

print(f"Número de registros com Zona de Conzumo 1 acima do limiar de 75%: {num_registros_acima_limite}")
print(f"Percentual da amostra que esses registros representam: {percentual_registros_acima_limite:.2f}%")

Número de registros com Zona de Conzumo 1 acima do limiar de 75%: 2388
Percentual da amostra que esses registros representam: 30.37%


In [ ]:
# 6. Calcule a temperatura média e crie um segundo DataFrame contendo consumo acima do limiar e temperatura acima da média.
media_temperatura = dados['Temperature'].mean()
print(f"A temperatura média é: {media_temperatura:.2f}")

df_consumo_temperatura_acima_media = dados[(dados['Consumo_Zona_1'] > d70) & (dados['Temperature'] > media_temperatura)]
display(df_consumo_temperatura_acima_media.head())

A temperatura média é: 18.79


,DateTime,Consumo_Zona_1,Consumo_Zona_2,Consumo_Zona_3,Temperature,Humidity,Wind Speed
12,6/5/2017 23:30,40993.90728,25383.4,24960.0,21.03,86.50,0.068
14,8/23/2017 17:50,37475.16093,24276.2,25140.6,27.06,47.58,4.905
19,6/22/2017 2:40,40676.02649,24754.7,25095.9,19.20,65.81,4.924
21,8/23/2017 13:50,40256.07103,27416.3,25453.5,27.11,43.35,4.906
25,7/12/2017 9:50,36511.89369,25906.3,25654.9,26.92,70.20,4.909


#### 7.	Compare os dois conjuntos e explique o que ocorreu com a quantidade de registros após a inclusão da condição ambiental.

In [ ]:
num_registros_consumo_temperatura = df_consumo_temperatura_acima_media.shape[0]
percentual_consumo_temperatura = (num_registros_consumo_temperatura / dados.shape[0]) * 100

print(f"Número de registros com Consumo_Zona_1 acima do limiar e Temperatura acima da média: {num_registros_consumo_temperatura}")
print(f"Percentual da amostra que esses registros representam: {percentual_consumo_temperatura:.2f}%")

Número de registros com Consumo_Zona_1 acima do limiar e Temperatura acima da média: 1597
Percentual da amostra que esses registros representam: 20.31%


Comparando os dois conjuntos de registros:

*   **`df70` (Consumo_Zona_1 alto):** Contém **2388 registros**, representando **30.37%** da amostra total.
*   **`df_consumo_temperatura_acima_media` (Consumo_Zona_1 alto E Temperatura acima da média):** Contém **1597 registros**, representando **20.31%** da amostra total.

### Efeito da Inclusão da Condição Ambiental (Temperatura)

A inclusão da condição ambiental (temperatura acima da média) como um segundo critério teve o efeito de **reduzir significativamente o número de registros selecionados**. Isso indica que:

1.  **Refinamento da Seleção:** Ao adicionar a temperatura como critério, estamos isolando um subconjunto mais específico dos eventos de alto consumo. Nem todos os períodos de alto consumo na Zona 1 ocorrem quando a temperatura ambiente está acima da média. Existe uma parcela de alto consumo que acontece sob temperaturas médias ou abaixo da média.
2.  **Identificação de Padrões Específicos:** Este novo conjunto de 1597 registros representa os momentos em que a alta demanda de energia na Zona 1 é *diretamente associada* a condições de temperatura mais elevadas. Isso pode ser crucial para a equipe de planejamento de energia, pois sugere que o uso de aparelhos de refrigeração, ar condicionado ou outros equipamentos sensíveis à temperatura pode ser um fator contribuinte significativo para esses picos de consumo.

Compreender essa correlação permite à equipe focar seus esforços em estratégias de gestão da demanda que considerem as condições climáticas, como programas de eficiência energética ou incentivos para reduzir o consumo durante ondas de calor.

## Dataset 4 — Solar Power Generation Data (Kaggle)

fonte: https://www.kaggle.com/datasets/anikannal/solar-power-generation-data?resource=download

Situação
Uma equipe de operação de uma usina fotovoltaica deseja localizar períodos de alta geração e identificar quais inversores aparecem com maior frequência nesses momentos. Nesta atividade, o foco inicial será o arquivo de geração da planta, sem realizar junção com o arquivo de sensores.


In [ ]:
# 1.	Carregue a amostra e renomeie as principais variáveis para nomes mais claros, como Potencia_CC, Potencia_CA e Geracao_Diaria.
dados = pd.read_csv('Solar_power_sample.csv')
dados.rename(columns={'DC_POWER': 'Potencia_CC'}, inplace=True)
dados.rename(columns={'AC_POWER': 'Potencia_CA'}, inplace=True)
dados.rename(columns={'DAILY_YIELD': 'Geracao_Diaria'}, inplace=True)

dados.head()

,DATE_TIME,SOURCE_KEY,Potencia_CC,Potencia_CA,Geracao_Diaria,TOTAL_YIELD
0,10-06-2020 19:00,uHbuxQJl8lW7ozc,0.00,0.000,6645.000,7240508.000
1,28-05-2020 00:15,rGa61gmuvPhdLxV,0.00,0.000,0.000,7206549.000
2,17-05-2020 15:15,sjndEbLyjtCKgGv,7333.88,718.062,6442.000,7036055.000
3,01-06-2020 04:15,ih0vzX44oOqAx2f,0.00,0.000,0.000,6308232.000
4,17-05-2020 09:00,z9Y9gH1T5YWrNuG,7665.50,750.438,961.375,7021701.375


In [ ]:
#2.	Apresente head(), shape, info() e describe().
dados.head()

,DATE_TIME,SOURCE_KEY,Potencia_CC,Potencia_CA,Geracao_Diaria,TOTAL_YIELD
0,10-06-2020 19:00,uHbuxQJl8lW7ozc,0.00,0.000,6645.000,7240508.000
1,28-05-2020 00:15,rGa61gmuvPhdLxV,0.00,0.000,0.000,7206549.000
2,17-05-2020 15:15,sjndEbLyjtCKgGv,7333.88,718.062,6442.000,7036055.000
3,01-06-2020 04:15,ih0vzX44oOqAx2f,0.00,0.000,0.000,6308232.000
4,17-05-2020 09:00,z9Y9gH1T5YWrNuG,7665.50,750.438,961.375,7021701.375


In [ ]:
dados.describe()

,Potencia_CC,Potencia_CA,Geracao_Diaria,TOTAL_YIELD
count,13756.000000,13756.000000,13756.000000,1.375600e+04
mean,3135.848662,306.695495,3319.235216,6.978829e+06
std,4023.069619,393.124284,3142.249927,4.162164e+05
min,0.000000,0.000000,0.000000,6.183645e+06
25%,0.000000,0.000000,0.000000,6.509779e+06
50%,370.071500,35.771400,2730.425000,7.146218e+06
75%,6350.562500,622.065500,6293.782500,7.267996e+06
max,14416.100000,1405.590000,9163.000000,7.846821e+06


In [ ]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13756 entries, 0 to 13755
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   DATE_TIME       13756 non-null  object 
 1   SOURCE_KEY      13756 non-null  object 
 2   Potencia_CC     13756 non-null  float64
 3   Potencia_CA     13756 non-null  float64
 4   Geracao_Diaria  13756 non-null  float64
 5   TOTAL_YIELD     13756 non-null  float64
dtypes: float64(4), object(2)
memory usage: 644.9+ KB


In [ ]:
dados.shape

(13756, 6)

In [ ]:
#3.	Determine a maior potência CA registrada.
dmax = dados['Potencia_CA'].max()
print(f"A maior potência CA registrada é: {dmax:.3f}")

A maior potência CA registrada é: 1405.590


In [ ]:
#4.	Calcule um limiar correspondente a 70% da potência CA máxima.
d70 = dmax*0.7
print(f"O limiar máximo é: {d70:.3f}")

O limiar máximo é: 983.913


In [ ]:
# 5.	Crie um DataFrame contendo somente registros acima desse limite e calcule sua quantidade e percentual.
df70 = dados[dados['Potencia_CA'] > d70]
df70.head()

,DATE_TIME,SOURCE_KEY,Potencia_CC,Potencia_CA,Geracao_Diaria,TOTAL_YIELD
11,26-05-2020 12:15,bvBOhCH3iADSZry,11830.1,1153.83,4028.14,6396364.143
46,15-06-2020 14:30,wCURE6d3bPkepu2,10571.3,1032.07,5366.57,7015064.571
48,25-05-2020 10:15,z9Y9gH1T5YWrNuG,11050.1,1079.86,2174.29,7082812.286
72,13-06-2020 11:30,7JYdWkrLSPkdwr4,11094.5,1082.44,3570.25,7815987.250
76,07-06-2020 13:00,VHMLBKoKgIrUVDU,12663.4,1234.69,4976.71,7382965.714


In [ ]:
# 6.	Utilize value_counts() em SOURCE_KEY dentro do DataFrame de alta geração.
df70['SOURCE_KEY'].value_counts()

,count
SOURCE_KEY,
adLQvlD726eNBSB,68
McdE0feGgRqW7Ca,65
ZnxXDlPa8U1GXgE,63
pkci93gMrogZuBj,62
ZoEaEvLYb1n2sOq,61
1IF53ai7Xc0U56Y,59
YxYtjZvoooNbGkE,58
sjndEbLyjtCKgGv,57
z9Y9gH1T5YWrNuG,56


#### 7.	Identifique qual inversor aparece com maior frequência nos registros de alta geração e descreva o que esse resultado permite observar, sem concluir desempenho ou falha apenas com esse recorte.



In [ ]:
inversor_mais_frequente = df70['SOURCE_KEY'].value_counts().idxmax()
contagem_inversor_mais_frequente = df70['SOURCE_KEY'].value_counts().max()

print(f"O inversor que aparece com maior frequência nos registros de alta geração é: {inversor_mais_frequente}")
print(f"Ele aparece {contagem_inversor_mais_frequente} vezes neste subconjunto de dados.")

O inversor que aparece com maior frequência nos registros de alta geração é: adLQvlD726eNBSB
Ele aparece 68 vezes neste subconjunto de dados.


A identificação de que o inversor **`adLQvlD726eNBSB`** (ou qualquer outro que seja o mais frequente) aparece mais vezes nos registros de alta geração indica que ele esteve **operando em alta capacidade** mais frequentemente do que os outros inversores durante os períodos considerados de alta geração. Isso não permite concluir diretamente sobre o seu desempenho superior ou qualquer falha. Pode ser que este inversor:

*   **Está localizado em uma área com maior exposição solar**;
*   **Possui maior capacidade instalada** em comparação com outros inversores, o que naturalmente o levaria a gerar mais energia e, consequentemente, a atingir o limiar de alta geração com mais frequência;
*   **Estava em operação por mais tempo** ou de forma mais consistente durante os períodos de pico de geração solar.

Para tirar conclusões sobre desempenho ou falha, seria necessário analisar métricas adicionais, como a eficiência de cada inversor, o tempo de atividade total, a ocorrência de manutenções ou interrupções, e comparar a geração em relação à sua capacidade nominal e às condições ambientais locais.

## Dataset 5 — Wind & Solar Energy Production (Kaggle)

fonte: https://www.kaggle.com/datasets/ahmeduzaki/wind-and-solar-energy-production-dataset

Situação
Um operador de um portfólio de fontes renováveis deseja comparar a ocorrência de períodos de alta produção solar e alta produção eólica. Como as duas fontes possuem escalas próprias, cada uma deverá ser comparada com o seu próprio valor máximo.


In [ ]:
# 1.	Renomeie as variáveis principais para Geracao_Solar e Geracao_Eolica.
dados = pd.read_csv('Wind_power_sample.csv')
dados.rename(columns={'Source': 'Recurso'}, inplace=True)
dados.rename(columns={'Production': 'Produção'}, inplace=True)

dados.head()

,Date,Start_Hour,End_Hour,Produção,Recurso,Day_of_Year,Day_Name,Month_Name,Season
0,12/18/2022,21,22,11400,Wind,352,Sunday,December,Winter
1,1/29/2024,14,15,7917,Wind,29,Monday,January,Winter
2,8/1/2024,15,16,8835,Solar,214,Thursday,August,Summer
3,11/10/2020,1,2,989,Wind,315,Tuesday,November,Fall
4,12/19/2022,1,2,12526,Wind,353,Monday,December,Winter


In [ ]:
# 2.	Determine o valor máximo registrado para Recurso.
dados['Produção'].max()


22929

In [ ]:
# 3.	Calcule separadamente 70% do máximo solar e 70% do máximo eólico.
dmax_solar = dados[dados['Recurso'] == 'Solar']['Produção'].max()
dmax_eolico = dados[dados['Recurso'] == 'Wind']['Produção'].max()


In [ ]:
# 4.	Crie um DataFrame para registros de alta geração solar e outro para registros de alta geração eólica.
df70_solar = dados[(dados['Recurso'] == 'Solar') & (dados['Produção'] > dmax_solar*0.7)]
df70_eolico = dados[(dados['Recurso'] == 'Wind') & (dados['Produção'] > dmax_eolico*0.7)]
display(df70_solar.head())
display(df70_eolico.head())

,Date,Start_Hour,End_Hour,Produção,Recurso,Day_of_Year,Day_Name,Month_Name,Season
77,7/8/2025,13,14,13125,Solar,189,Tuesday,July,Summer
178,3/19/2025,13,14,13786,Solar,78,Wednesday,March,Spring
183,8/17/2025,12,13,11455,Solar,229,Sunday,August,Summer
187,7/16/2025,12,13,16160,Solar,197,Wednesday,July,Summer
207,3/31/2025,12,13,12441,Solar,90,Monday,March,Spring


,Date,Start_Hour,End_Hour,Produção,Recurso,Day_of_Year,Day_Name,Month_Name,Season
20,1/3/2024,0,1,17272,Wind,3,Wednesday,January,Winter
32,3/11/2021,13,14,16144,Wind,70,Thursday,March,Spring
103,1/29/2025,0,1,18199,Wind,29,Wednesday,January,Winter
151,12/8/2024,6,7,16946,Wind,343,Sunday,December,Winter
192,11/19/2024,2,3,16568,Wind,324,Tuesday,November,Fall


In [ ]:
# 5.	Conte os registros de cada DataFrame e calcule seus respectivos percentuais.
num_registros_acima_limite_solar = df70_solar.shape[0]
percentual_registros_acima_limite_solar = (num_registros_acima_limite_solar / dados.shape[0]) * 100

num_registros_acima_limite_eolico = df70_eolico.shape[0]
percentual_registros_acima_limite_eolico = (num_registros_acima_limite_eolico / dados.shape[0]) * 100

In [ ]:
# 6.	Compare qual fonte aparece com maior frequência acima de 70% do seu próprio máximo.
if num_registros_acima_limite_solar > num_registros_acima_limite_eolico:
    print(f"A fonte 'Solar' aparece com maior frequência acima de 70% do seu próprio máximo.")
    print(f"Número de registros acima de 70% do máximo solar: {num_registros_acima_limite_solar}")
    print(f"Percentual da amostra que esses registros representam: {percentual_registros_acima_limite_solar:.2f}%")
else:
    print(f"A fonte 'Wind' aparece com maior frequência acima de 70% do seu próprio máximo.")
    print(f"Número de registros acima de 70% do máximo eólico: {num_registros_acima_limite_eolico}")
    print(f"Percentual da amostra que esses registros representam: {percentual_registros_acima_limite_eolico:.2f}%")


A fonte 'Wind' aparece com maior frequência acima de 70% do seu próprio máximo.
Número de registros acima de 70% do máximo eólico: 245
Percentual da amostra que esses registros representam: 2.36%


#### 7.	Explique por que não seria adequado utilizar diretamente o mesmo valor numérico de potência como limite para as duas fontes sem antes observar a escala de cada variável.

#### Explicação sobre a inadequação de um mesmo limite numérico para diferentes fontes de energia

Não seria adequado utilizar diretamente o mesmo valor numérico de potência (por exemplo, 10.000 kW) como limite para as duas fontes de energia (solar e eólica) sem antes observar a escala de cada variável por algumas razões fundamentais:

1.  **Diferenças na Capacidade Instalada e no Potencial de Geração:** A capacidade máxima de geração de uma usina solar pode ser significativamente diferente da de uma usina eólica. Cada tipo de fonte possui características intrínsecas e limites tecnológicos que determinam o seu potencial máximo de produção. Usar um limite absoluto que se baseia, por exemplo, no máximo da fonte eólica, pode fazer com que a fonte solar nunca o atinja, ou vice-versa, distorcendo a análise de 'alta geração'.

2.  **Variações na Natureza da Geração:** A geração solar depende diretamente da irradiação solar, que varia com a hora do dia, estação do ano e condições climáticas. A geração eólica depende da velocidade do vento, que também possui padrões de variabilidade distintos. Essas diferentes naturezas de geração significam que seus perfis de produção e picos são intrinsecamente diferentes.

3.  **Comparabilidade Distorcida:** Se aplicarmos um limite de 15.000 kW, por exemplo, e a usina solar tiver um máximo histórico de 12.000 kW, ela nunca atingiria o critério de 'alta geração', mesmo quando estivesse operando em sua capacidade máxima. Por outro lado, se a usina eólica tivesse um máximo histórico de 25.000 kW, 15.000 kW poderia ser considerado uma produção 'média' para ela. Isso levaria a conclusões errôneas ou comparações injustas entre as fontes.

4.  **Objetivos da Análise:** Ao comparar a 'ocorrência de períodos de alta produção', o que se busca é entender quando cada fonte está operando em um regime próximo ao seu potencial máximo ou em uma capacidade significativamente elevada *para ela mesma*. Definir um percentual do máximo individual (como 70% ou 75% do seu próprio valor máximo) permite uma análise mais justa e relevante, destacando os períodos de alta performance relativa para cada tecnologia, independentemente de suas escalas absolutas. Isso foi feito nas etapas anteriores ao calcular `dmax_solar` e `dmax_eolico` separadamente e então aplicar um percentual a cada um.

## Dataset 6 — Individual Household Electric Power Consumption (UCI)

fonte: https://archive.ics.uci.edu/dataset/235/individual+household+electric+power+consumption

Situação
Uma residência possui monitoramento elétrico detalhado e deseja identificar episódios de demanda elevada que também apresentem corrente acima do comportamento médio. O dataset é o mesmo utilizado como referência em aula, mas o critério de análise será diferente.


In [ ]:
# 1.	Carregue a nova amostra e organize os atributos utilizando nomes mais simples.
dados = pd.read_csv('Individual_household_sample.csv')
dados.rename(columns={'Global_active_power': 'Potencia_Ativa_Global'}, inplace=True)
dados.rename(columns={'Global_reactive_power': 'Potencia_Reactiva_Global'}, inplace=True)
dados.rename(columns={'Voltage': 'Tensão'}, inplace=True)
dados.rename(columns={'Global_intensity': 'Intensidade_Global'}, inplace=True)
dados.rename(columns={'Sub_metering_1': 'Potencia_Sub_1'}, inplace=True)
dados.rename(columns={'Sub_metering_2': 'Potencia_Sub_2'}, inplace=True)
dados.rename(columns={'Sub_metering_3': 'Potencia_Sub_3'}, inplace=True)

dados.head()

,Potencia_Ativa_Global,Potencia_Reactiva_Global,Tensão,Intensidade_Global,Potencia_Sub_1,Potencia_Sub_2,Potencia_Sub_3
0,1.502,0.074,240.17,6.4,0,0,18
1,0.374,0.264,245.50,1.8,0,2,0
2,0.620,0.300,239.85,3.0,0,1,1
3,0.280,0.200,235.72,1.4,0,0,0
4,1.372,0.054,243.95,5.6,0,0,18


In [ ]:
# 2.	Determine o valor máximo de potência ativa.
dados['Potencia_Ativa_Global'].max()

10.536

In [ ]:
# 3.	Calcule 75% do valor máximo e crie um DataFrame contendo registros acima desse limite.
dmax = dados['Potencia_Ativa_Global'].max()
df75 = dmax*0.75
df75 = dados[dados['Potencia_Ativa_Global'] > df75]
df75.head()

,Potencia_Ativa_Global,Potencia_Reactiva_Global,Tensão,Intensidade_Global,Potencia_Sub_1,Potencia_Sub_2,Potencia_Sub_3
3132,8.540,0.238,236.23,36.0,80,35,17
3630,8.278,0.088,231.50,35.6,0,71,17
10339,8.026,0.066,236.45,33.8,37,73,18
11695,8.102,0.190,229.31,35.4,35,35,17
27983,10.290,0.302,230.90,44.6,35,66,17


In [ ]:
# 4.	Calcule a quantidade e o percentual de registros selecionados.
num_registros_acima_limite = df75.shape[0]
percentual_registros_acima_limite = (num_registros_acima_limite / dados.shape[0]) * 100

print(f"Número de registros com Potencia_Ativa_Global acima do limiar de 75%: {num_registros_acima_limite}")
print(f"Percentual da amostra que esses registros representam: {percentual_registros_acima_limite:.4f}")


Número de registros com Potencia_Ativa_Global acima do limiar de 75%: 58
Percentual da amostra que esses registros representam: 0.0283


In [ ]:
# 5.	Calcule a corrente média da amostra.
media_corrente = dados['Intensidade_Global'].mean()
print(f"A corrente média da amostra é: {media_corrente:.2f}")

A corrente média da amostra é: 4.63


In [ ]:
# 6.  Crie um segundo DataFrame contendo simultaneamente potência ativa acima de 75% do máximo e corrente acima da média.
limiar_potencia_ativa_alta = dados['Potencia_Ativa_Global'].max() * 0.75
df_consumo_corrente_acima_media = dados[(dados['Potencia_Ativa_Global'] > limiar_potencia_ativa_alta) & (dados['Intensidade_Global'] > media_corrente)]
display(df_consumo_corrente_acima_media.head())

,Potencia_Ativa_Global,Potencia_Reactiva_Global,Tensão,Intensidade_Global,Potencia_Sub_1,Potencia_Sub_2,Potencia_Sub_3
3132,8.540,0.238,236.23,36.0,80,35,17
3630,8.278,0.088,231.50,35.6,0,71,17
10339,8.026,0.066,236.45,33.8,37,73,18
11695,8.102,0.190,229.31,35.4,35,35,17
27983,10.290,0.302,230.90,44.6,35,66,17


#### 7.	Compare os dois conjuntos e explique o efeito da inclusão da corrente como segunda condição.

In [ ]:
num_registros_duplo_criterio_corrente = df_consumo_corrente_acima_media.shape[0]
percentual_duplo_criterio_corrente = (num_registros_duplo_criterio_corrente / dados.shape[0]) * 100

print(f"Número de registros com Potencia_Ativa_Global acima do limiar de 75% E Intensidade_Global acima da média: {num_registros_duplo_criterio_corrente}")
print(f"Percentual da amostra que esses registros representam (com duplo critério): {percentual_duplo_criterio_corrente:.4f}%")

Número de registros com Potencia_Ativa_Global acima do limiar de 75% E Intensidade_Global acima da média: 58
Percentual da amostra que esses registros representam (com duplo critério): 0.0283%


Comparando os dois conjuntos de registros:

*   **`df75` (apenas consumo ativo alto):** Contém **58 registros**, representando **0.0283%** da amostra total.
*   **`df_consumo_corrente_acima_media` (consumo ativo alto E corrente acima da média):** Contém **58 registros**, representando **0.0283%** da amostra total.

### Efeito da Inclusão da Condição Ambiental (Corrente)

Neste caso específico, a inclusão da condição de `Intensidade_Global` acima da média **não alterou o número de registros selecionados** em relação ao filtro apenas por `Potencia_Ativa_Global` acima de 75% do máximo. Isso sugere que, para este dataset e para o critério de alta potência ativa definido:

1.  **Correlação Forte:** Os eventos de `Potencia_Ativa_Global` muito alta estão altamente correlacionados com `Intensidade_Global` acima da média. Ou seja, sempre que a potência ativa excede 75% do seu máximo, a corrente também já está acima da média.
2.  **Redundância Parcial do Critério:** A segunda condição (`Intensidade_Global` > `media_corrente`) não está excluindo registros adicionais porque o conjunto de registros que já atende ao primeiro critério (`Potencia_Ativa_Global` > 75% do máximo) já está, em sua totalidade, atendendo também ao segundo critério.

Para a equipe de monitoramento, isso pode indicar que a `Intensidade_Global` acima da média não atua como um filtro adicional para identificar um subconjunto mais específico de eventos de interesse quando a potência ativa já está em níveis muito elevados. Em outras palavras, um *alto consumo de eletrodomésticos* (potência ativa) já é um bom indicativo de *alta corrente* neste cenário.